stl/inc/xutility

```diff
@@ -4198,7 +4198,368 @@ _NODISCARD _CONSTEXPR17 move_iterator<_Iter> make_move_iterator(_Iter _It) { //
     return move_iterator<_Iter>(_STD move(_It));
 }
 
+// FUNCTION TEMPLATE _Construct_in_place
+template <class _Ty, class... _Types>
+void _Construct_in_place(_Ty& _Obj, _Types&&... _Args) noexcept(is_nothrow_constructible_v<_Ty, _Types...>) {
+    ::new (const_cast<void*>(static_cast<const volatile void*>(_STD addressof(_Obj))))
+        _Ty(_STD forward<_Types>(_Args)...);
+}
+
+#if _HAS_CXX17
+// STRUCT _Nontrivial_dummy_type
+struct _Nontrivial_dummy_type {
+    constexpr _Nontrivial_dummy_type() noexcept {
+        // This default constructor is user-provided to avoid zero-initialization when objects are value-initialized.
+    }
+};
+_STL_INTERNAL_STATIC_ASSERT(!is_trivially_default_constructible_v<_Nontrivial_dummy_type>);
+#endif // _HAS_CXX17
+
 #ifdef __cpp_lib_concepts
+// CLASS common_iterator
+// clang-format off
+template <input_or_output_iterator _Iter, sentinel_for<_Iter> _Se>
+    requires (!same_as<_Iter, _Se> && copyable<_Iter>)
+class common_iterator {
+    // clang-format on
+    class _Proxy {
+        iter_value_t<_Iter> _Keep;
+
+    public:
+        _Proxy(iter_reference_t<_Iter>&& _Right) : _Keep(_STD move(_Right)) {}
+
+        const iter_value_t<_Iter>* operator->() const {
+            return _STD addressof(_Keep);
+        }
+    };
+
+public:
+    constexpr common_iterator() noexcept(is_nothrow_constructible_v<_Iter>) // strengthened
+        : _Myiter{}, _Mystate{_State::_Holds_iter} {};
+
+    constexpr common_iterator(_Iter _Right) noexcept(is_nothrow_move_constructible_v<_Iter>) // strengthened
+        : _Myiter{_STD move(_Right)}, _Mystate{_State::_Holds_iter} {}
+
+    constexpr common_iterator(_Se _Right) noexcept(is_nothrow_move_constructible_v<_Se>) // strengthened
+        : _Mysentinel{_STD move(_Right)}, _Mystate{_State::_Holds_sentinel} {}
+
+    constexpr common_iterator(const common_iterator& _Right) noexcept(
+        is_nothrow_copy_constructible_v<const _Iter&>&& is_nothrow_copy_constructible_v<const _Se&>) // strengthened
+        : _Mydummy{}, _Mystate{_Right._Mystate} {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Mystate != _State::_Nothing, "common_iterator can only be assigned if it holds a value");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        if (_Mystate == _State::_Holds_iter) {
+            _Construct_in_place(_Myiter, _Right._Myiter);
+        } else if (_Mystate == _State::_Holds_sentinel) {
+            _Construct_in_place(_Mysentinel, _Right._Mysentinel);
+        }
+    }

```

Comment:
According to the standard there are indeed preconditions for the copy [constructor](http://eel.is/c++draft/iterators.common#common.iter.const-3)/ [assignment](http://eel.is/c++draft/iterators.common#common.iter.const-5)

So the first point is not applicable.

The standard explicitely defines only copy constructor / copy assignment so I believe it would be invalid to add move constructor / move assignment. Maybe one of the maintainers can confirm.

I cannot follow your thoughts on triviality. Could you elaborate?

Reply:
Those are converting constructor/assignment. AFAIK, there should be implicitly defined copy/move constructor/assignment (thus inheriting triviality from `variant`).

Reply:
The implicitly defined copy / move constructor is deleted due to the union. I initially tried it without them and the compiler bit me hard

Reply:
Under the as if rule, we should implement copy/move constructor/assignment with the same semantics as the implicit ones of the wording.

Reply:
Under the as if rule, we should implement copy/move constructor/assignment with the same semantics as the implicit ones of the wording.

Reply:
I think the rule would be that if a copy constructor is defined there is no [implicit ](http://eel.is/c++draft/class.mem#class.copy.ctor-8)move constructor.

Reply:
Copy constructor is not declared in the wording. (`common_iterator(const common_iterator<I2, S2>&)` isn’t)

Reply:
I have to think about this more, I was assuming that this counts as the copy constructor for `I2==I && S2==S` but indeed a copy constructor is a [non template function](http://eel.is/c++draft/class.mem#class.copy.ctor-1)